app/helpers/application_helper.rb

```diff
@@ -52,4 +52,73 @@ def sample_types_page_expermient?
       @my_module.nil? &&
       !@experiment.nil?
   end
+
+  def smart_annotation_parser(text)
+    sa_reg = /\[\#(.*?)~(prj|exp|tsk|sam)~([0-9]+)\]/
+    new_text = text.gsub(sa_reg) do |el|
+      match = el.match(sa_reg)
+      case match[2]
+      when 'prj'
+        project = Project.find_by_id(match[3].to_i)
+        next unless project
+        if project.archived?
+          "<span class='sa-type'>#{sanitize(match[2])}</span> " \
+          "#{link_to project.name,
+                     projects_archive_path} #{t'atwho.res.archived'}"
+        else
+          "<span class='sa-type'>#{sanitize(match[2])}</span> " \
+          "#{link_to project.name,
+                     project_path(project)}"
+        end
+      when 'exp'
+        experiment = Experiment.find_by_id(match[3].to_i)
+        next unless experiment
+        if experiment.archived?
+          "<span class='sa-type'>#{sanitize(match[2])}</span> " \
+          "#{link_to experiment.name,
+                     experiment_archive_project_path(experiment.project)} " \
+          "#{t'atwho.res.archived'}"
+        else
+          "<span class='sa-type'>#{sanitize(match[2])}</span> " \
+          "#{link_to experiment.name,
+                     canvas_experiment_path(experiment)}"
+        end
+      when 'tsk'
+        my_module = MyModule.find_by_id(match[3].to_i)
+        next unless my_module
+        if my_module.archived?
+          "<span class='sa-type'>#{sanitize(match[2])}</span> " \
+          "#{link_to my_module.name,
+                     module_archive_experiment_path(my_module.experiment)}" \
+          "#{t'atwho.res.archived'}"
+        else
+          "<span class='sa-type'>#{sanitize(match[2])}</span> " \
+          "#{link_to my_module.name,
+                     protocols_my_module_path(my_module)}"
+        end
+      when 'sam'
+        sample = Sample.find_by_id(match[3])
+        if sample
+          "<span class='glyphicon glyphicon-tint'></span> " \
+          "#{link_to sample.name,
+                     samples_project_path(sample
+                                          .organization
+                                          .projects
+                                          .first)}"
+        else
+          "<span class='glyphicon glyphicon-tint'></span> " \
+          "#{match[1]} #{t'atwho.res.deleted'}"
+        end
+      end
+    end
+
+    sa_user = /\[\@(.*?)~([0-9]+)\]/
```

Comment:
This Regex is wrong, because it doesn't cover Base62 (Base62 is `0-9a-zA-Z`, I'd imagine) for IDs.